src/NuGet.Core/NuGet.ProjectModel/Utf8JsonStreamReader.cs

```diff
@@ -0,0 +1,338 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Buffers;
+using System.Collections.Generic;
+using System.IO;
+using System.Linq;
+using System.Text;
+using System.Text.Json;
+
+namespace NuGet.ProjectModel
+{
+    /// <summary>
+    /// This struct is used to read over a memeory stream in parts, in order to avoid reading the entire stream into memory.
+    /// It functions as a wrapper around <see cref="Utf8JsonStreamReader"/>, while maintaining a stream and a buffer to read from.
+    /// must call <see cref="CompleteRead"/> to return the buffer to the pool when completed.
+    /// </summary>
+    internal ref struct Utf8JsonStreamReader
+    {
+        private static readonly char[] DelimitedStringDelimiters = new char[] { ' ', ',' };
+
+        // The UT8 BOM characters which may appear at the start of a JSON file.
+        private ReadOnlySpan<byte> _utf8Bom = new byte[] { 0xEF, 0xBB, 0xBF };
+        private Utf8JsonReader _reader;
+        // The buffer is used to read from the stream in chunks.
+        private byte[] _buffer;
+        // The stream is the source of the JSON data.
+        private Stream _stream;
+        private bool _complete;
+
+        internal Utf8JsonStreamReader(Stream stream)
+        {
+            if (stream is null)
+            {
+                throw new ArgumentNullException(nameof(stream));
+            }
+
+            _complete = false;
+            _stream = stream;
+            _buffer = ArrayPool<byte>.Shared.Rent(1024);
+            _stream.Read(_buffer, 0, 3);
+            var offset = 0;
+            if (!_utf8Bom.SequenceEqual(_buffer.AsSpan(0, 3)))
```

Comment:
Do we need to worry about this?  We could just write the file without the BOM to avoid this.  Of course if one is already on disk and people upgrade NuGet, it could fail...

Reply:
that's the question, I added it as a precaution but I have no idea if it's actually needed.

Reply:
It doesn't look like any of the asset files I have on my computer contain a BOM. 

I rather keep the check, that way if a go to parse a different file we don't risk having it break because we don't check for BOM.